File: week6/929 - Number Maze.cpp
#include<stdio.h>
#include<queue>
#include<vector>

using namespace std;

int matrix [1000][1000];
int dj [1000][1000];
int m,n;

typedef pair<int, int> pos;
typedef pair<int, pos> elm;
priority_queue<elm, vector<elm>, greater<elm> > q;

void init(){
    for(int i =0 ;i < m; i++){
        for (int j = 0; j < n; j++){
            dj[i][j] = 10000000;

        }
    }
}



int main(){
    int T;
    scanf("%d",&T);
    while(T--){
        scanf("%d",&m);
        scanf("%d",&n);
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                scanf("%d",&matrix[i][j]);
            }
        }
        init();

        q.push(make_pair(matrix[0][0],pos(0,0)));
        dj[0][0] = matrix[0][0];
        //printf("%d,%d", q.top().first,q.top().second.first);
        while(true){
            elm atual = q.top();
            q.pop();
            int posX = atual.second.first;
            int posY = atual.second.second;
            if(posX > 0){
                if(dj[posX][posY]+matrix[posX-1][posY] < dj[posX-1][posY]){
                    dj[posX-1][posY] = dj[posX][posY]+matrix[posX-1][posY];
                    q.push(make_pair(matrix[posX-1][posY]+dj[posX][posY],pos(posX-1,posY)));
                }
            }

            if(posY > 0){
                if(dj[posX][posY]+matrix[posX][posY-1] < dj[posX][posY-1]){
                    dj[posX][posY-1] = dj[posX][posY]+matrix[posX][posY-1];
                    q.push(make_pair(matrix[posX][posY-1]+dj[posX][posY],pos(posX,posY-1)));
                }
            }
            if(posX < m-1){
                if(dj[posX][posY]+matrix[posX+1][posY] < dj[posX+1][posY]){
                    dj[posX+1][posY] = dj[posX][posY]+matrix[posX+1][posY];
                    q.push(make_pair(matrix[posX+1][posY]+dj[posX][posY],pos(posX+1,posY)));
                }
            }

            if(posY < n-1){
                if(dj[posX][posY]+matrix[posX][posY+1] < dj[posX][posY+1]){
                    dj[posX][posY+1] = dj[posX][posY]+matrix[posX][posY+1];
                    q.push(make_pair(matrix[posX][posY+1]+dj[posX][posY],pos(posX,posY+1)));
                }
            }
            if(q.empty()) break;


        }
        printf("%d\n",dj[m-1][n-1]);

    }


    return 0;
}

File: week6/929 - Number Maze.py
from queue import PriorityQueue
from sys import stdin

class MyPriorityQueue(PriorityQueue):
    def __init__(self):
        PriorityQueue.__init__(self)
        self.counter = 0

    def put(self, item, priority):
        PriorityQueue.put(self, (priority, self.counter, item))
        self.counter += 1

    def get(self, *args, **kwargs):
        _, _, item = PriorityQueue.get(self, *args, **kwargs)
        return item

def main():
        T = int(stdin.readline())

        for t in range(T):
                q = MyPriorityQueue()
                m = int(stdin.readline())
                n = int(stdin.readline())
                matrix = [[int(x) for x in stdin.readline().split()]for y in range(m)]
                dj = [[1000000 for x in range(n)]for y in range(m)]
        
                #dijkstra
                dj[0][0] = matrix[0][0]
                posX = 0
                posY = 0
                q.put((0,0),0)
                while(True):
                        if(posX > 0):
                                if(dj[posX][posY]+matrix[posX-1][posY] < dj[posX-1][posY]):
                                        dj[posX-1][posY] = dj[posX][posY]+matrix[posX-1][posY]
                                        q.put((posX-1,posY),dj[posX][posY]+matrix[posX-1][posY])
                                                
                        if(posY > 0):
                                if(dj[posX][posY]+matrix[posX][posY-1] < dj[posX][posY-1]):
                                        dj[posX][posY-1] = dj[posX][posY]+matrix[posX][posY-1]
                                        q.put((posX,posY-1),dj[posX][posY]+matrix[posX][posY-1])
                                     
                        if(posX < m-1):
                                if(dj[posX][posY]+matrix[posX+1][posY] < dj[posX+1][posY]):
                                        dj[posX+1][posY] = dj[posX][posY]+matrix[posX+1][posY]
                                        q.put((posX+1,posY),dj[posX][posY]+matrix[posX+1][posY])
                                
                        if(posY < n-1):
                                if(dj[posX][posY]+matrix[posX][posY+1] < dj[posX][posY+1]):
                                        dj[posX][posY+1] = dj[posX][posY]+matrix[posX][posY+1]
                                        q.put((posX,posY+1),dj[posX][posY]+matrix[posX][posY+1])
                                
                        posX,posY = q.get()
                        
                        if(q.empty()):
                                break
                print(dj[m-1][n-1])
main()
